scripts/rosdistro_freeze_source

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
+
+from rosdistro.index import Index
+
+import argparse
+import os.path
+import subprocess
+import sys
+import threading
+import time
+import yaml
+
+try:
+    import queue
+except ImportError:
+    import Queue as queue
+
+from rosdistro import get_distribution
+from rosdistro.writer import yaml_from_distribution_file
+
+CONCURRENT_OPERATIONS=16
+
+
+def parse_args(args=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description='Freeze a rosdistro\'s source branch versions to hashes or tags.'
+    )
+    parser.add_argument('index',
+        help='Path to a local index.yaml file.'
+    )
+    parser.add_argument('dist_names', nargs='*',
+        help='The names of the distributions (default: all)'
+    )
+    parser.add_argument('--release-version', action="store_true", default=False,
+        help='Freeze to the hash of current release tag (default: the source branch hash)'
+    )
+    return parser.parse_args(args)
+
+
+def freeze_distribution_sources(dist, release_version=False):
+    # Populate this queue with tuples of repositories instances to be updated,
+    # so that this work can be spread acros multiple threads.
+    work_queue = queue.Queue()
+    for repo_name, repo in dist.repositories.iteritems():
+        if repo.source_repository:
+            version = repo.source_repository.version
+            if release_version and repo.release_repository:
+                version = repo.release_repository.version.split('-')[0]
+            work_queue.put((repo.source_repository, version))
+
+    total_items = work_queue.qsize()
+
+    for i in range(CONCURRENT_OPERATIONS):
+        threading.Thread(target=_worker, args=[work_queue]).start()
+
+    # Wait until the threads have done all the work and exited.
+    while not work_queue.empty():
+        time.sleep(0.1)
+        sys.stdout.write("Updating source repo versions (%d/%d)   \r" %
+                         (total_items - work_queue.qsize(), total_items))
+        sys.stdout.flush()
+    work_queue.join()
+
+    # Clear past the updating line.
+    print("")
+
+
+def _worker(work_queue):
+    while True:
+        try:
+            source_repo, freeze_version = work_queue.get(block=False)
+            cmd = ['git', 'ls-remote', source_repo.url]
+            ls_remote_lines = subprocess.check_output(cmd).splitlines()
+            for line in ls_remote_lines:
+                hash, ref = line.split('\t', 1)
+                if ref.endswith(freeze_version):
+                    source_repo.version = hash
+                    break
+
+            work_queue.task_done()
+
+        except subprocess.CalledProcessError:
+            print("Non-zero return code for: %s" % ' '.join(cmd))
+            work_queue.task_done()
+
+        except queue.Empty:
+            break
+
+
+def main():
+    args = parse_args()
+    with open(args.index) as f:
+       index = Index(yaml.safe_load(f), 'file://' + os.path.dirname(os.path.abspath(args.index)))
```

Comment:
There are more robust ways to do it with urlparse etc but it's effectively the same. Of course if you want to do it python2/3 it's going to need to be conditional ;-) https://stackoverflow.com/questions/11687478/convert-a-filename-to-a-file-url